// @private/models/BookStore.model.ts
import * as mongoose from 'mongoose';
import { BookModelInterface } from '@private/models/Book.model';
import { StoreModelInterface } from '@private/models/Store.model';

const schema = new mongoose.Schema({
    book: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'book',
        index: true
    },
    store: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'store',
        index: true
    },
    is_active: {type: Boolean, default: true},
    amount: { type: Number, default: 0 },
    quantity_sold: { type: Number, default: 0 }
}, {
    timestamps: {
        createdAt: 'created_at',
        updatedAt: 'updated_at'
    }
});

export interface BookStoreModelInterface extends mongoose.Document {
    book: mongoose.Types.ObjectId | BookModelInterface,
    store: mongoose.Types.ObjectId | StoreModelInterface,
    amount: number
    is_active: boolean
    quantity_sold: number
    created_at: number
    updated_at: number
}

export default mongoose.model<BookStoreModelInterface>('book_store', schema, 'book_store');
