// @private/models/Purchase.model.ts
import * as mongoose from 'mongoose';
import User, { UserModelInterface } from '@private/models/User.model';
import Book, { BookModelInterface } from '@private/models/Book.model';
import ShippingCost, { ShippingCostModelInterface } from '@private/models/ShippingCost.model';

const statics = {
    PENDING: 'PENDING',
    DONE: 'DONE'
};

const schema = new mongoose.Schema({
    user: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'user',
        index: true
    },
    book: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'book',
        index: true
    },
    status: { type: String, default: statics.PENDING },
    number: { type: Number, default: 0 },
    shipping: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'shipping_cost',
        index: true
    },
    amount: { type: Number, default: 0 }
}, {
    timestamps: {
        createdAt: 'created_at',
        updatedAt: 'updated_at'
    }
});

schema.pre('save', async function (next) {
    let err;
    let that: any = this;

    let book_data = await Book.findOne({ _id: that.book, is_active: true });
    if (!book_data) {
        err = new Error('Book not found');
        next(err);
    }

    let user_data = await User.findOne({ _id: that.user, is_active: true });
    if (!user_data) {
        err = new Error('User not found');
        next(err);
    }

    let shipping_cost_data = await ShippingCost.findOne({ _id: that.shipping, is_active: true });
    if (!shipping_cost_data) {
        err = new Error('Shipping cost not found');
        next(err);
    }


});

export interface PurchaseModelInterface extends mongoose.Document {
    user: mongoose.Types.ObjectId | UserModelInterface
    book: mongoose.Types.ObjectId | BookModelInterface
    status: string
    number: number
    shipping: mongoose.Types.ObjectId | ShippingCostModelInterface
    amount: number
    created_at: Date,
    updated_at: Date
}

interface PurchaseStaticsInterface extends mongoose.Model<PurchaseModelInterface> {
    PurchaseModelInterface: PurchaseModelInterface
}

export default mongoose.model<PurchaseModelInterface, PurchaseStaticsInterface>('purchase', schema, 'purchase');
