// @private/models/Staff.model.ts
import * as mongoose from 'mongoose';
import Store, { StoreModelInterface } from '@private/models/Store.model';

const statics = {
    MALE: 'MALE',
    FEMALE: 'FEMALE'
};

const schema = new mongoose.Schema({
    user_name: { type: String, unique: true },
    name: String,
    name_slug: { type: String, index: true },
    date_of_birth: { type: Number, default: 0 },
    email: { type: String, unique: true },
    sex: { type: String, default: statics.MALE },
    id_card_number: { type: String, default: '' },
    id_card_number_date: { type: Number, default: 0 },
    id_card_number_location: { type: String, default: '' },
    tax_number: { type: String, default: '' },
    insurrance_number: { type: String, default: '' },
    start_work_date: { type: Number, default: 0 },
    end_work_date: { type: Number, default: 0 },
    salt: String,
    password: String,
    phone_number: {type: String, index: true},
    address: String,
    store: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'store',
        index: true
    },
    is_active: { type: Boolean, default: true },
    last_login: Number,
    role: String
}, {
    timestamps: {
        createdAt: 'created_at',
        updatedAt: 'updated_at'
    }
});

schema.virtual('id').get(function () {
    return this._id.toHexString();
});


schema.pre('save', async function (next) {
    let err;
    let that: any = this;

    let store_data = await Store.findOne({ _id: that.store });
    if (!store_data) {
        err = new Error('Store not found');
        next(err);
    }


});


export interface StaffModelInterface extends mongoose.Document {
    user_name: string
    email: string
    name: string
    name_slug: string
    date_of_birth: number
    sex: string
    id_card_number: string
    id_card_number_date: number
    id_card_number_location: string
    tax_number: string
    store: mongoose.Types.ObjectId | StoreModelInterface
    insurrance_number: string
    start_work_date: number
    end_work_date: number
    salt: string
    password: string
    phone_number: string
    address: string
    role: string
    is_active: boolean
    last_login: number
    created_at: Date
    updated_at: Date
}

export default mongoose.model<StaffModelInterface>('staff', schema, 'staff');
