// @private/configs/index.ts
declare var process: any;
export const SERVER_VERSION = '4.0.0';


export const JWT_SECRET                    = process.env.JWT_SECRET;
export const MONGODB_URI                   = process.env.MONGODB_URI;
export const PORT                          = process.env.PORT;
export const GRAPHQL_PORT                  = process.env.GRAPHQL_PORT;
export const SECRET_KEY_MACHINE            = process.env.SECRET_KEY_MACHINE;
export const QRCODE_TTL                    = process.env.QRCODE_TTL || 60;
export const ADMIN_EMAIL                   = process.env.ADMIN_EMAIL;
export const EMAIL                         = process.env.EMAIL;
export const PASS                          = process.env.PASS;
export const CACHE_TTL                     = process.env.CACHE_TTL || 60;
export const CACHE_OTP                     = process.env.CACHE_OTP || 300;
export const FCM_SERVER_KEY                = process.env.FCM_SERVER_KEY;
export const LIMIT_MOBILE_REQUEST_PER_HOUR = process.env.LIMIT_MOBILE_REQUEST_PER_HOUR;
export const PREFIXED                      = process.env.PREFIX;

export const REDIS = {
    HOST: process.env.REDIS_HOST,
    PORT: process.env.REDIS_PORT,
    PREFIX: process.env.REDIS_PREFIX,
    PASSWORD: process.env.REDIS_PASSWORD
};

export const HASH_PASSWORD       = process.env.HASH_PASSWORD;
export const FIREBASE_AUD        = process.env.FIREBASE_AUD;
export const TIMEZONE            = process.env.TIMEZONE;
export const SERVER_ORGANIZATION = process.env.SERVER_ORGANIZATION;
export const TOKEN_WEB_SALT      = process.env.TOKEN_WEB_SALT;
export const TOKEN_CMS_SALT      = process.env.TOKEN_CMS_SALT;

export const ACCESS_TOKEN_CMS_EXPIRE_IN      = process.env.ACCESS_TOKEN_CMS_EXPIRE_IN;
export const ACCESS_TOKEN_WEB_EXPIRE_IN   = process.env.ACCESS_TOKEN_WEB_EXPIRE_IN;
export const REFRESH_TOKEN_CMS_EXPIRE_IN     = process.env.REFRESH_TOKEN_CMS_EXPIRE_IN;
export const REFRESH_TOKEN_WEB_EXPIRE_IN  = process.env.REFRESH_TOKEN_WEB_EXPIRE_IN;

export const EXCEPTION_REPORT = [
    'Exception',
    'ValidationError'
];

export const CACHE = {
    REDIS_USER: 'USER_'
};
